File: legacy/app/services/route_finder.rb
require "uri"
require "net/http"

class RouteFinder < ApplicationService
  attr_reader :end_location
  API_URL = "https://api.external.citymapper.com/api/1/directions/transit?"

  def initialize(start_location, end_location)
    @end_location = end_location
    @start_location = start_location
  end

  def call
    params = {
      start: @start_location,
      end: @end_location,
      time: Time.now.iso8601,
      time_type: 'depart',
      language: 'en'
    }
    url = URI(API_URL+params.to_param)

    https = Net::HTTP.new(url.host, url.port)
    https.use_ssl = true

    request = Net::HTTP::Get.new(url, {'Citymapper-Partner-Key' => ENV["CITY_MAPPER_KEY"]})


    response = https.request(request)
    route = Route.new
    route.from_json(response.read_body)

    return route
  end
end

File: lib/route-finder.ts
import { Route } from './models';

export class RouteFinder {
  private startLocation: string;
  private endLocation: string;

  constructor(startLocation: string, endLocation: string) {
    this.startLocation = startLocation;
    this.endLocation = endLocation;
  }

  async call(): Promise<Route> {
    const apiUrl =
      process.env.CITYMAPPER_API_URL ||
      'https://api.external.citymapper.com/api/1/directions/transit?';

    const params = new URLSearchParams({
      start: this.startLocation,
      end: this.endLocation,
      time: new Date().toISOString(),
      time_type: 'depart',
      language: 'en',
    });

    const url = `${apiUrl}${params.toString()}`;
    const response = await fetch(url, {
      headers: {
        'Citymapper-Partner-Key': process.env.CITY_MAPPER_KEY || '',
      },
    });

    if (!response.ok) {
      throw new Error(`Citymapper API error: ${response.statusText}`);
    }

    const route: Route = await response.json();
    return route;
  }

  static async call(
    startLocation: string,
    endLocation: string
  ): Promise<Route> {
    const finder = new RouteFinder(startLocation, endLocation);
    return finder.call();
  }
}

